[ENT-334] Add client-side registration form validation.

UmanShahzad opened the pull request:
## Front-end

### Focus on a field (input/textarea/select ONLY)

![selection_003](https://user-images.githubusercontent.com/10018065/28566942-28361286-714a-11e7-9b79-5a05e5b0983d.png)

### Blur and get a SUCCESS validation (input/textarea/select)

![selection_002](https://user-images.githubusercontent.com/10018065/28566789-a1505998-7149-11e7-83e2-b6566fe18bce.png)

### Check and get a SUCCESS validation (checkbox)

![selection_006](https://user-images.githubusercontent.com/10018065/28566808-b5602b98-7149-11e7-948b-820d227c04c2.png)

### Focus on SUCCESS validation (input/textarea/select ONLY)

![selection_004](https://user-images.githubusercontent.com/10018065/28566922-101466ee-714a-11e7-84b4-8f8adf6cef38.png)

### SUCCESS validation fades (after 3 seconds) (input/textarea/select)

![selection_005](https://user-images.githubusercontent.com/10018065/28566876-e4c0b75e-7149-11e7-9088-6535d9aca2ce.png)

### SUCCESS validation fades (after 3 seconds) (checkbox)

![selection_007](https://user-images.githubusercontent.com/10018065/28566859-d88e464a-7149-11e7-9664-22da328495a2.png)

### Blur and get an ERROR validation (input/textarea/select)

![selection_001](https://user-images.githubusercontent.com/10018065/28566693-5e8e3a76-7149-11e7-87d0-c08aacbc32c7.png)

### Uncheck and get an ERROR validation (checkbox)

![image](https://user-images.githubusercontent.com/10018065/28566671-506a3896-7149-11e7-8aef-274e826d56a1.png)


## Back-end
This implements a validation API for registration, where a client makes AJAX calls to the endpoints at

```bash
/api/user/v1/validation/registration
```

requesting validation decisions on each input. Inputs are of the form

```json
{
    "name": "Uman Shahzad",
    "email": "edx@email.com",
    "confirm_email": "edx@email.com",
    "username": "myusername",
    "password": "mypassword",
    "country": "PK"
}
```

However, a user may just as well input

```json
{
    "email": "edx@email.com"
}
```

to get a validation decision for just the `email` field alone. The same applies for all other form fields for which there exists a handler (see below for more info on handlers).

Responses come in JSON format and provide strings for each input form field dependent on the type of validation error; if no error, then empty string to indicate OK. For example, in response to the latter input above, which is a valid email:

```json
{
    "validation_decisions": {
        "email": ""
    }
}
```

If the email had any issues in its format, we'd get an appropriate error message response in place of `""`. Clients can then use this error message directly on the interface with styling.

See documentation for more detail on what we can input & get back.

### Handlers

A handler in this API's architecture is just a method. These handlers simply return an error message string, or an empty string if no error occurred in validation. There currently exist implementations of handlers for 

- `name`
- `email`
- `confirm_email`
- `password`
- `username`
- `country`

Extending the API for more handlers is as simple as adding a method for handling, that takes in a `request` and outputs an error message (or an empty string), then adding this method to the dictionary of available validation handlers.

## Other

**JIRA tickets**: Implements ENT-334's back-end version (front-end to come).

**Discussions**: The architecture of the validation API is covered extensively in the `RegistrationValidationView`'s documentation, and above.

**Dependencies**: None

**Screenshots**: Back-end only right now -- screenshots to come when front-end work is up (next week).

**Sandbox URL**: https://pr15375.sandbox.opencraft.hosting/

**Deployment targets**: edx.org and edge.edx.org

**Merge deadline**: Hopefully by start of July.

**Testing instructions**:

1. Go to the registration page with this branch running on your devstack or the sandbox (see if URL is available above).
1. Input into any field proper and improper inputs for that field, click somewhere else or tab, and see if it appropriately notifies you right away that there is an issue or that you entered correctly.

**Author notes and concerns**:

1. The `RegistrationValidationView`'s handlers could use some refactoring, potentially -- instead of individual methods, we can make these instances of classes like `PasswordValidation`, etc., but I believe this may complicate things unnecessarily because the method-handler point of view is in itself still perfectly fine.
2. The cross/check FA icons are visualized on the requirements as being on the right side of the input field. However, this could require a large amount of CSS code to change, because of the way the registration container is made, and potentially break mobile or simply not work on that platform either way (because input fields take up full horizontal space on small devices). I propose a different way of handling the icons:

![](http://i.imgur.com/6nb2BYm.png)
and
![](http://i.imgur.com/lhZGGWh.png)

Having them right next to the label is clear, obvious, viewable on mobile without mobile-specific changes, and a cake to code up (I already did). It would also be possible to do this cleanly for other fields which don't take up full horizontal space (like dropdown lists).

**Reviewers**
- [ ] @haikuginger 
- [ ] @brittneyexline @douglashall @chris-mike @cptvitamin 

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
EDXAPP_REGISTRATION_EXTRA_FIELDS:
  confirm_email: "required"
  level_of_education: "required"
  gender: "required"
  year_of_birth: "required"
  mailing_address: "required"
  goals: "required"
  honor_code: "required"
  terms_of_service: "required"
  city: "required"
  country: "required"
```

edx-webhook commented:
Thanks for the pull request, @UmanShahzad! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=15375&repo=edx%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

UmanShahzad commented on lms/djangoapps/student_account/test/test_views.py:
```diff
@@ -62,24 +62,6 @@ class StudentAccountUpdateTest(CacheIsolationTestCase, UrlResetMixin):
-        u"{user}@example.com".format(
-            user=(u'e' * (EMAIL_MAX_LENGTH - 11))
-        )
-    ]
```
Just repeat code sitting around.

UmanShahzad commented on lms/djangoapps/student_account/test/test_views.py:
```diff
@@ -31,7 +31,7 @@
 from commerce.tests.mocks import mock_get_orders
 from course_modes.models import CourseMode
-from edxmako.shortcuts import render_to_response
+from openedx.core.djangoapps.user_api.accounts.tests.testutils import INVALID_EMAILS
```
I'll have to sort this down in the list in the next commit.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -783,10 +782,6 @@ def _validate_password(password, username=None):
-    try:
-        _validate_unicode(password)
-    except UnicodeError as invalid_unicode_err:
-        raise AccountPasswordInvalid(str(invalid_unicode_err))
```
Didn't realize passwords aren't actually checked for unicode issues.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -240,20 +243,6 @@ def update_account_settings(requesting_user, update, username=None):
-
-    existing_user_profile, _ = UserProfile.objects.get_or_create(user=existing_user)
-
-    return existing_user, existing_user_profile
```
Just reorganization to put "semi-private" functions together at the bottom.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -416,6 +408,169 @@ def request_password_change(email, orig_host, is_secure):
+        _validate_username_doesnt_exist(username)
+    except (AccountUsernameAlreadyExists):
+        return True
+    return False
```
I'm not currently actually using the `check_X_is_valid` or `check_X_exists` functions, but I wrote them earlier because I felt I had needed them, when in fact I didn't later. Maybe it'll come in handy down the road for someone else.

brittneyexline replied:
I generally feel like we shouldn't keep functions that aren't being used around - it could cause confusion for other developers down the line.

UmanShahzad replied:
It could, but developer's IDEs nowadays have a feature to check for usage. I have a feeling that a function like this *could* come in handy.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -416,6 +408,169 @@ def request_password_change(email, orig_host, is_secure):
+        _validate_username_doesnt_exist(username)
+    except AccountUsernameAlreadyExists as username_exists_err:
+        return str(username_exists_err)
+    return default
```
I wanted to make existence checks separate because it's not always expected that one gets a response based off of existence when checking for whether the string is "valid".

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -429,32 +584,189 @@ def _validate_username(username):
+    :raises: AccountUsernameInvalid
+    """
+    _validate_username_satisfies_min_length(username)
+    _validate_username_satisfies_max_length(username)
```
Maybe a little overkill in breaking down the functions, but it's honestly way more clear.

haikuginger replied:
I disagree. At this point, we've got six different functions to handle checking the length of a string. Perhaps consolidate to one function that takes a string, a max length, and a min length, then call it from each of `validate_username_of_good_length`, `_validate_email_of_good_length` and `_validate_password_of_good_length`? Then, we'd have small, cohesive blocks of logic. Each `_validate` function could pull the relevant settings and pass them in, the single length-checking function can make its check, and the calling function can output the relevant error message if needed.

UmanShahzad replied:
I fold there -- I had the same thought as you, but the real reason I didn't end up abstracting it this way is because I wanted to keep the actual exception class that's raised dependent on the error. But I can just scratch the various `AccountXInvalid`, `AccountXAlreadyExists` type exceptions and squash them into a more generic one and then solve this particular problem generically too.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -471,62 +783,91 @@ def _validate_password(password, username):
+    try:
+        _validate_unicode(password)
+    except UnicodeError as invalid_unicode_err:
+        raise AccountPasswordInvalid(str(invalid_unicode_err))
```
Removed unicode check for passwords in next commit.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/tests/test_api.py:
```diff
@@ -294,46 +307,12 @@ class AccountCreationActivationAndPasswordChangeTest(TestCase):
     """
     USERNAME = u'frank-underwood'
-    PASSWORD = u'ṕáśśẃőŕd'
+    PASSWORD = u'password'
```
Reverted these next commit.

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/tests/test_api.py:
```diff
@@ -294,46 +307,12 @@ class AccountCreationActivationAndPasswordChangeTest(TestCase):
-        u'',
-        u'a',
-        u'a' * (PASSWORD_MAX_LENGTH + 1)
-    ]
```
These were being reused so dropped them into a utilities file.

UmanShahzad commented on openedx/core/djangoapps/user_api/legacy_urls.py:
```diff
@@ -38,10 +38,19 @@
+            r'^v1/account/password_reset/$',
+            user_api_views.PasswordResetView.as_view(),
+            name="user_api_password_reset"
+        ),
```
I think we should keep this style rollin'.

haikuginger replied:
As below. If you're not changing it as part of necessary work, don't change it.

UmanShahzad commented on openedx/core/djangoapps/user_api/tests/test_views.py:
```diff
@@ -31,7 +31,6 @@
 from .test_helpers import TestCaseForm
 from xmodule.modulestore.tests.django_utils import SharedModuleStoreTestCase
 from xmodule.modulestore.tests.factories import CourseFactory
-from ..helpers import FormDescription
```
An unused artifact of the SAML country code PR :)

UmanShahzad commented on openedx/core/djangoapps/user_api/validation/tests/test_views.py:
```diff
@@ -0,0 +1,191 @@
+            self.assertNotEqual(
+                self.get_validation_decision({form_field_name: data}),
+                {form_field_name: ''}
+            )
```
Does `ddt` provide something here for me to avoid the for loop? How would I iterate over each invalid email with `email` as the `form_field_name` separately, and so on with just `ddt`?

haikuginger replied:
@UmanShahzad, I believe you would want to break the test into three - one for emails, one for passwords, and one for usernames - then, you could DDT each one. Or, you could do something like this:

```
@ddt.data(
    [('email', item) for item in INVALID_FIELDS[1:]] +
    [('password', item) for item in INVALID_PASSWORDS[1:]] +
    [('username', item) for item in INVALID_USERNAMES[1:]]
)
@ddt.unpack
def test_negative_validation_decision(self, form_field_name, field_value):
    pass
```

UmanShahzad replied:
@haikuginger I had these as 3 tests, and tried to simplify by doing it this way. I think I'll follow the style you've posted.

UmanShahzad commented on openedx/core/lib/api/test_utils.py:
```diff
@@ -28,7 +28,7 @@ def request_with_auth(self, method, *args, **kwargs):
 
     def get_json(self, *args, **kwargs):
-        """Make a request with the given args and return the parsed JSON repsonse"""
+        """Make a request with the given args and return the parsed JSON response"""
```
=)

haikuginger commented on openedx/core/djangoapps/user_api/validation/views.py:
```diff
@@ -0,0 +1,145 @@
+        **Example Requests and Responses**
+
+            - Checks the validity of the username and email inputs separately.
+            GET /api/user/v1/validation/registration/
```
GET seems wrong; let's go with POST.

brittneyexline replied:
Agreed.

UmanShahzad replied:
Woops! Yeah on second thought, it does sound wrong.

haikuginger commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -347,10 +336,13 @@ def check_account_exists(username=None, email=None):
     conflicts = []
 
-    if email is not None and User.objects.filter(email=email).exists():
+    try:
```
I don't see the need to incur the overhead of an exception handler on this call that already works fine.

UmanShahzad replied:
I had to reuse this functionality elsewhere so I decided to drop it into a separate function, and decided to reuse it here.

haikuginger commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -471,62 +782,87 @@ def _validate_password(password, username):
+            raise UnicodeError
+        # In some cases we pass the above, but it's still inappropriate utf-8.
+        str(data)
+    except UnicodeError:
```
We seem to be doing a lot of Unicode checking that we didn't do in the previous version. What led to this change? Were we failing to consider some edge cases properly?

UmanShahzad replied:
According to the lists we had in tests that showed invalid emails/usernames, there were ones that only had unicode-related problems. But in my own testing when I used those lists I would get errors related to those strings in the lists, so I attempted to solve that here.

haikuginger commented on openedx/core/djangoapps/user_api/validation/views.py:
```diff
@@ -0,0 +1,145 @@
+                validation_decisions.update({
+                    form_field_key: handler(self, request)
+                })
+        return Response({"validation_decisions": validation_decisions})
```
We should be returning with an error type (400-range) if changes need to be made, 200 otherwise. This will simplify the success-case Javascript substantially.

brittneyexline replied:
Yea, I also agree with this point. Though I'm wondering if that means each of these handlers would then become their own views, since it would be ambiguous to return 400 with a request that had one valid input and one invalid input.

UmanShahzad replied:
The architecture of this API was meant to work so that, indeed, a client could put in multiple inputs at once and get responses for each individual input in one go, and at the same time get a response for an input that may have depended on others in the same request.

Changing this to returning numerical error types could lead to the issue @brittneyexline mentioned, and I really don't think we should be turning these into multiple views -- a simple, flexible, reusable back-end is what I was preferring for this PR.

haikuginger replied:
We can still return bodies on an error code, but I'll defer to you on this. I suspect that the current architecture might make the front-end Javascript work a bit more difficult, but you can of course make the backend change later on if you find that to be the case.

haikuginger commented on openedx/core/djangoapps/user_api/urls.py:
```diff
@@ -25,9 +26,21 @@
-    url(r'^v1/me$', ME, name='own_username_api'),
-    url(r'^v1/accounts/{}$'.format(settings.USERNAME_PATTERN), ACCOUNT_DETAIL, name='accounts_api'),
-    url(r'^v1/accounts$', ACCOUNT_LIST, name='accounts_detail_api'),
+    url(
```
Please don't change code unrelated to what you're doing, even if the style is less than ideal.

UmanShahzad replied:
I agree with the `legacy_urls` file, but here I was editing it to be aligned with the style right below it, a style I decided to continue with the addition of `RegistrationValidationView`'s endpoint.

haikuginger commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -471,62 +782,87 @@ def _validate_password(password, username):
+        if not isinstance(data, str) and not isinstance(data, unicode):
+            raise UnicodeError
+        # In some cases we pass the above, but it's still inappropriate utf-8.
+        str(data)
```
What are we trying to accomplish here? I see what it does, but I don't understand _why_.

UmanShahzad replied:
I was running into situations where a string like the one in one of the files below in the diffs (`u'invalid-unicode_fŕáńḱ'`) was giving me a lot of errors in tests. It passes the instance checks so it was flying by this validation but then giving errors later. The `str(data)` will ensure this error happens for these kinds of situations so I can pick it up. I could be way off here.

haikuginger replied:
Okay; let's leave that as-is for now, then.

brittneyexline commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -471,62 +782,87 @@ def _validate_password(password, username):
+        # In some cases we pass the above, but it's still inappropriate utf-8.
+        str(data)
+    except UnicodeError:
+        raise UnicodeError(u"Input not valid unicode")
```
it appears that we're raising the exception that we're catching right after and then raising this, why not just raise it in the above clause?

UmanShahzad replied:
In the first case that we raise it, it's for the instance check failure. Then we call `str(data)` to see if for some reason the string passed the instance checks but was still invalid unicode (like mentioned above and seen through the list of invalid username/emails from the tests). If it was invalid there, it'll raise some kind of UnicodeError right there and we catch it. So with this catch block we catch both cases of the UnicodeError but output the same error message string.

UmanShahzad commented:
@haikuginger @brittneyexline Thanks for your reviews :) I believe the requested changes aren't that substantial so I'll try to get something up that'll satisfy an approval for Monday. I'm pretty sure we'll be ready to start the front-end for next week.

douglashall commented:
@UmanShahzad I have cherry-picked this commit onto a sprint branch (edx/edx-platform:mighty_mouse) and deployed that branch to https://business.sandbox.edx.org.

UmanShahzad commented:
@douglashall Nice! Thanks! It seems to be working:

`https://business.sandbox.edx.org/api/user/v1/validation/registration?username=`
`{"validation_decisions":{"username":"Username '' must be at least 2 characters long"}}`

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/__init__.py:
```diff
@@ -25,3 +26,13 @@
+
+# Translators: This message is shown to users who attempt to create a new
+# account using an email address associated with an existing account.
+EMAIL_CONFLICT_MSG = u"It looks like {email_address} belongs to an existing account. " \
```
I also planned on adding more strings here of the sort, to reuse these in tests (I've been through too many situations now where the error string is hardcoded into tests; it's a little annoying to have to deal with if it changes). But I'd still like to know if there's simply no harm here in translating before being used.

UmanShahzad replied:
(deleted my old comments)

Yep, I had this before, but erased it because I thought to separate out the message itself from when it's being translated. But I think now it's okay to just pass it around already with the translation function called on it.

UmanShahzad commented:
@haikuginger Thanks :)

UmanShahzad commented:
@brittneyexline @haikuginger Working on front-end now. It will be in this same PR.

UmanShahzad commented:
@brittneyexline @haikuginger @douglashall The latest commit implements a prototype of the front-end, which is now only missing displaying the actual error string. Validation occurs on blur and input forms & their labels are properly changing colors, etc.. I have not touched any other existing validation code lest I break something.

I need some recommendations on where exactly the error message should be displayed. Right below the input forms but above any existing instructions? Or maybe something more fancier?

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -158,6 +177,18 @@
+                },
+
+                liveValidate: function(event) {
+                    var url = '/api/user/v1/validation/registration',
```
Any way to cleanly keep this in underscore templates?

UmanShahzad commented:
Ok I only just now got access to the screenshots for the UI portion. I'll try to get these up soon.

UmanShahzad commented:
Okay, this does (almost) everything as part of the requirements, except the part about showing the `(required)` string in `$red` only when focusing a required field, about checking for confirmation email matching email field, and the ARIA related requirements.

The latter 2 are simple enough, 1st should also be. I'd like to write the spec after it's fully featured, then we can start a code review. Try this on your devstack if you like.

UmanShahzad commented:
@brittneyexline @douglashall I have 2 questions:

1. Should success indicators disappear on refocus?
2. The cross/check FA icons are visualized on the requirements as being on the right side of the input field. However, this could require a large amount of CSS code to change, because of the way the registration container is made, and potentially break mobile or simply not work on that platform either way (because input fields take up full horizontal space on small devices). I propose a different way of handling the icons:

![](http://i.imgur.com/V61de9Q.png)

Having them right next to the label is clear, obvious, viewable on mobile without mobile-specific changes, and a cake to code up (I already did). What do you think about this?

P.S. If you want to try it on your devstack, please do, because all features (except validation for confirmation email errors and ARIA) are up.

(When the latter 2 features are up, I'll ping you guys again :) ).

UmanShahzad commented:
@haikuginger @brittneyexline @douglashall I believe this is ready for review, but I'd definitely need an answer for some of the previous questions (related to UI/UX).

itsjeyd commented on openedx/core/djangoapps/user_api/accounts/__init__.py:
```diff
@@ -25,3 +28,53 @@
+
+# Translators: This message is shown to users who attempt to create a new account using
+# an invalid email format.
+EMAIL_INVALID_MSG = _(u"Email '{email}' format is not valid.")
```
@UmanShahzad CC @brittneyexline @douglashall I tested this PR on my local machine and came across this validation message. When viewed in context (with an actual email address inserted for the placeholder) it sounds a little awkward (to me). For example:

> Email 'a@b@c' format is not valid.

Can we change this to something like:

> "a@b@c" is not a valid email address.

(Also note the double quotes.)

UmanShahzad replied:
@itsjeyd I definitely was looking for input on this -- thanks a lot. I found it awkward too, and I'd love it if you can comment on some of the other error messages as well. :+1: I like your suggestion, and will use that, unless anyone has another.

UmanShahzad commented:
Test failures are ephemeral -- I'll push a new commit with updated error messages according to @itsjeyd 's suggestion. Please do suggest fixes to them.

UmanShahzad commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -13,45 +21,47 @@
     <% if ( type === 'select' ) { %>
-        <select id="<%= form %>-<%= name %>"
-            name="<%= name %>"
+        <select id="<%- form %>-<%- name %>"
```
These are not spurious -- they were required to pass all the linting issues. (We need to use `-` instead of `=` here).

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -4,162 +4,293 @@
-        function($, _, gettext, FormView, formStatusTpl) {
-            return FormView.extend({
-                el: '#register-form',
+    ], function(
```
@haikuginger Yes, sure.

UmanShahzad commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -354,7 +358,11 @@
 
             &.error {
-                border-color: $error-color;
+                border-color: $red;
```
It's a part of the requirements (`$red` matches the requirements):

![](http://i.imgur.com/2nZiAy1.png)

![](http://i.imgur.com/0EtGdig.png)

UmanShahzad commented on lms/static/sass/views/_login-register.scss:
```diff
@@ -383,9 +391,14 @@
         }
+
+        .error, .label-required {
+            color: $red;
```
It's for the requirements.

UmanShahzad commented on lms/static/js/student_account/views/FormView.js:
```diff
@@ -201,24 +209,49 @@
+                    }
+                },
+
+                errorExists: function(id) {
```
Yep, I didn't catch this. Thanks!

UmanShahzad commented on lms/static/js/student_account/views/FormView.js:
```diff
@@ -201,24 +209,49 @@
+                },
+
+                doOnErrorList: function(id, action) {
+                    var i;
```
I _think_ it may have been an (awkward) linting issue. That's why past devs were declaring it earlier in the code. But I think it may not apply here -- will test it.

UmanShahzad commented on lms/static/js/student_account/views/FormView.js:
```diff
@@ -177,7 +177,9 @@
                             } else {
-                                errors.push(test.message);
+                                if (!this.errorExists(test.id)) {
+                                    errors.push(test.message);
```
Yeah I didn't like it either, changed it to `validation`.

UmanShahzad commented on lms/static/js/student_account/views/FormView.js:
```diff
@@ -177,7 +177,9 @@
                                 $label.removeClass('error');
                             } else {
-                                errors.push(test.message);
+                                if (!this.errorExists(test.id)) {
```
Right.

UmanShahzad commented on lms/static/js/student_account/views/FormView.js:
```diff
@@ -201,24 +209,49 @@
-                        scrollTop: this.$formFeedback.offset().top
-                    }, 'slow');
+                deleteError: function(id) {
+                    var self = this;
```
If I want to use a callback function and get access to `this.errors`, this is how, no pun intended.

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -4,162 +4,293 @@
+                }
+            },
+
+            renderLiveValidationError: function($el, $label, $icon, $tip, error) {
```
There is duplication, yeah, I'll try to see what I can do.

UmanShahzad commented:
@haikuginger Thanks for the review!

1. I'll look into why email validation for empty emails is coming back that way.
2. I'm not sure what the exact requirement is for the "(required)" text's color when we have success. I had asked a question that didn't get answered yet -- namely, whether we should remove success indicators on refocus.
3. Yep, I asked about this as well in a comment recently -- _should_ it go away? What's the workflow here exactly? I'm not sure I want to take guesses when someone knows (namely whoever's responsible for judging UX). I'm perfectly fine with making it always disappear on blur, but would like to get confirmation on whether that's the approach hoped for.

itsjeyd commented:
@UmanShahzad @haikuginger Regarding the UX details that are left to sort out, we discussed these at the demo meeting last night: Christine is not available this week, so can't weigh in. If we can reach consensus about the code, and about how the form behaves aside from the details that are left to sort out, we can merge this PR and address any follow-up feedback as part of a separate story.

CC @brittneyexline @douglashall

itsjeyd commented:
@UmanShahzad Small (but important) addendum to my previous [comment](https://github.com/edx/edx-platform/pull/15375#issuecomment-312256587): Since this isn't an Enterprise-specific change, you will get feedback about the current behavior from a couple more people, and it is possible that you'll have to make some more changes to the current behavior before we can merge this. I missed that somehow when we were discussing this PR last night.

CC @haikuginger @brittneyexline @douglashall

UmanShahzad commented:
@itsjeyd Sounds good to me, thanks for noting this. I'll get this stuff up soon as @haikuginger discussed thus far.

UmanShahzad commented:
@haikuginger As for the problem with the email, I tested it by sending a request to the edX's business sandbox:

```
(edxapp) edxapp@vagrant:~/edx-platform$ curl -X POST --data "email=" https://business.sandbox.edx.org/api/user/v1/validation/registration
{"validation_decisions":{"email":"Email must be between 3 and 254 characters long."}}
```

In other words I think we just _do_ in fact have an account without an email, most likely the superuser we create on our devstacks.

Should we leave it as be? Or make a special case for when the email is blank?

haikuginger commented:
@UmanShahzad, let's go with a special case for a blank email.

UmanShahzad commented:
@haikuginger Thanks! It should be as simple as changing this:

```python
return email_exists_error or invalid_email_error
```

to this

```python
return invalid_email_error or email_exists_error
```

UmanShahzad commented on lms/static/js/spec/student_account/register_spec.js:
```diff
@@ -143,10 +170,10 @@
+                                    {value: '', name: '--'},
+                                    {value: 1900, name: '1900'},
+                                    {value: 1950, name: '1950'},
+                                    {value: 2014, name: '2014'}
```
Hm, I don't remember even doing this. Think my IDE might have done it automatically. Aw well.

UmanShahzad commented:
@haikuginger Ready for review.

@brittneyexline @douglashall If you guys also have time for a code review, that'd be excellent :)

douglashall commented:
I have deployed the commits on this branch to business.sandbox.edx.org. It looks like we get form validation errors immediately upon loading the registration page. Also, @chris-mike had some comments on ENT-334(https://openedx.atlassian.net/browse/ENT-334) that we haven't addressed yet.

haikuginger commented:
@douglashall, I think the validation issues might be due to the sandbox being behind basic auth; it looks like the browser isn't wrapping the AJAX requests with an Authorization header. We're spinning up a sandbox on our own infrastructure for testing; @UmanShahzad believes that most of the outstanding mentioned issues are actually fixed in commits not deployed to the business sandbox.

douglashall commented:
If there is work from this branch missing from the business/nicholas_nickleby branch please merge this branch to business/nicholas_nickleby again and update the business sandbox. We want to test this in an integration testing environment and not on a one-off sandbox. Let me know if you need help resolving the basic auth issues you are seeing.

UmanShahzad commented:
@douglashall Getting form validation issues upon loading the page should only happen if your browser has form data that it wants to automatically fill out when you view a page with a form. In that case there's not much I can do here to prevent that, because the logic (as layed out in the requirements) works so that on blur should trigger validation. I in fact don't see this as a bad thing, because if the browser accidentally inputs `edxapp` into the `email` field, that sounds like something the user should see immediately for fixing.

As for Christine's comments, I think the only thing I haven't covered in the latest version of this is that the Country field (or any required field) should show "(required)" just as the others. That's something I'll begin work on now.

Finally, about updating the business sandbox -- I'm not sure I (personally) have access to do that. I also think that the features involved in this PR don't really require "integration" testing so to speak, because there's no dependency changes / touching many parts of the system (the backend and frontend are both in one PR). But either way, I can't update it so maybe @haikuginger or you can, whoever has time / access.

douglashall commented:
@UmanShahzad We'd like to reduce the overhead for testing ENT work, having a single integration testing sandbox allows us to avoid duplicating sandbox setup with every PR. I have given you SSH access to business.sandbox.edx.org. Let me know if you have trouble connecting.

chris-mike commented:
Some requested changes:

1. Change the X icon to  fa-exclamation-circle [&#xf06a;]. The X looks like close, rather than an error.
2. Have "(required)" appear in the same color/size as "(optional)". The red looks too much like an error before an error has occurred.
3. When the field is in an error state (wrong information, no information), the "required" indicator should then turn red. Color should be consistent with the state.
4. Can we not get icons to the right of the field? Do they have to be in-line with the field label?

douglashall commented:
@chris-mike What is the reasoning behind displaying the "(required)" text on focus of a field? If we are going to do that, why not just always displays the "(required)" text?

UmanShahzad commented:
@douglashall Ah, thanks! I've connected successfully, and will update the box soon, after addressing Christine's changes.

@chris-mike 

1. I can easily change it to `fa-exclamation-circle`, sure.
2. Sure, I had it the way it currently is because the requirements showed it that way. But what @douglashall said is also something to consider.
3. If we fix point 2, then yes, this would automatically be the case (i.e. red color on error, green color on success, matching the rest of the input box).
4. I wrote some things about why I believe we should go with this not being on the right side of input boxes in the PR summary, I'll copy it here:

The cross/check FA icons are visualized on the requirements as being on the right side of the input field. However, this could require a large amount of CSS code to change, because of the way the registration container is made, and potentially break mobile or simply not work on that platform either way (because input fields take up full horizontal space on small devices). I propose a different way of handling the icons:

![](http://i.imgur.com/6nb2BYm.png)
and
![](http://i.imgur.com/lhZGGWh.png)

Having them right next to the label is clear, obvious, viewable on mobile without mobile-specific changes, and a cake to code up (I already did). It would also be possible to do this cleanly for other fields which don't take up full horizontal space (like dropdown lists).

haikuginger commented:
@UmanShahzad, if you're updating the sandbox manually, be sure to rebuild the static assets. The process won't be exactly the same, but something like what's shown here:

https://github.com/open-craft/documentation/blob/master/sandboxes_howto.md

douglashall commented:
@haikuginger @UmanShahzad There are script aliases installed on the business sandbox which makes updating the code running on the sandbox a bit easier. I will send out a more comprehensive email to the team shortly, but the following would allow you to update edx-platform on the sandbox:
```
# Switch to edxapp user, activate edxapp virtualenv, cd to edx-platform dir
$ edxapp
# Pull latest business/nicholas_nickleby branch
$ git pull
# Install requirements, migrate, build static
$ edxapp_upgrade
# Exit the edxapp user and restart the edxapp service
$ edxapp_restart
```

UmanShahzad commented:
@haikuginger @douglashall Thanks guys! Those script aliases are much needed even for devstack installs...

I don't have write access to the repo, though, so I'll have to merge this branch from the sandbox and push the changes from there.

douglashall commented:
@UmanShahzad That should work. We should definitely get you write access to the edX repositories though. @itsjeyd @bradenmacdonald Is that something you could help Uman with?

UmanShahzad commented:
@douglashall OpenCraft doesn't go through the process of giving a trial developer access, until he's a core member, so that's why I don't have access yet.

P.S. I've finished all the features, but I have to be right back, then I'll update the box and we can see to it that this is ready to go (minus the accessibility review).

UmanShahzad commented:
@douglashall @haikuginger @chris-mike The changes are made and the sandbox updated -- except validation isn't happening due to:

`https://business.sandbox.edx.org/api/user/v1/validation/registration 403 (FORBIDDEN)`

haikuginger commented:
Okay, so I looked into the auth failure - what's going on is that the POST is being sent through directly to Django without passing through the basic auth layer for some reason, and so _Django_, rather than nginx, is getting that header to interpret. And, since that particular username/password combo for the sandbox doesn't exist, it's rejecting the request out-of-hand. Not sure how to handle this; will come back later.

douglashall commented:
@UmanShahzad @haikuginger I think the issue with authentication is that we shouldn't require authentication for the validation endpoint since it needs to be available to anonymous users. You can disable authentication for `RegistrationValidationView` by setting `authentication_classes` to an empty list, similar to what we do for `RegistrationView` [here](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/user_api/views.py#L159).

haikuginger commented:
@douglashall, thanks for the hint! I've updated the PR and the sandbox with that change, and it appears to be working correctly. CC @chris-mike.

UmanShahzad commented:
@chris-mike I've updated the business sandbox to reflect your change requests on the ticket. Let me know how it looks, now.

CC @haikuginger @douglashall @brittneyexline

itsjeyd commented:
@UmanShahzad Could you please point out the commits that you added recently, in case @haikuginger / @brittneyexline / @douglashall would like to review them individually?

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -4,29 +4,49 @@
+                // Select boxes are unfortunately special cases,
+                // because one "focuses" on the first item, not the input box.
+                    'blur select[required]:first': 'liveValidateHandler',
+                    'focus select[required]:first': 'handleRequiredInputFocus'
```
Validation for select fields will only happen after you blur from your selection of the list. Selecting an item does not unfortunately count as blurring from the list -- you have to explicitly select the item, then click away from the list.

As for which child to choose, as far as I've seen, looking for blur/focus events on the first child has been sufficient for no matter what child you select.

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -77,6 +98,130 @@
+
+                    // Hide success indicators after some time.
+                    if (this.liveValidationFields[name]) {
+                        clearTimeout(this.liveValidationFields[name]);
```
We could set it to `null` after clearing it up, sure. As for setting it to `null` in the timeout function, though, that would then make it impossible to clear the timeout as we're doing here because we'd lose the timeout's reference.

haikuginger replied:
I mean, set it to null inside the function that the timeout executes - so it would only be cleared out when that function executes (at the end of the timeout).

UmanShahzad replied:
Ahh, ok got it :)

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -133,32 +282,81 @@
                     }
 
+                    if ($confirmEmail.length) {
+                        if (!$confirmEmail.val() || $emailElement.val() !== $confirmEmail.val()) {
```
So I've been following a pattern (as it is in other places in the codebase and I believe with JS devs who use jquery in general) where anything prefixed with a `$` is a reference to an HTML element. So when I check for the length, I'm really seeing if the object that represents this element exists at all. Then I check for the value, which is an individual key-value pair in that object. This style is used in other places and yes, for JS in general I believe, to check for existence.

As for the order of operations, sure, I can put that last part into parantheses.

haikuginger replied:
Got it; makes sense.

UmanShahzad commented on lms/static/js/student_account/views/RegisterView.js:
```diff
@@ -133,32 +282,81 @@
+                    var $el = $(event.currentTarget),
+                        type = $el.attr('type');
+                    if (this.inValidationFields($el)) {
+                        if (type === 'checkbox') {
```
Sure, no problem!

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -20,16 +21,24 @@
+    EMAIL_MIN_LENGTH, PASSWORD_MIN_LENGTH, USERNAME_MIN_LENGTH,
+    EMAIL_MAX_LENGTH, PASSWORD_MAX_LENGTH, USERNAME_MAX_LENGTH,
+    PASSWORD_CANT_EQUAL_USERNAME_MSG, PASSWORD_EMPTY_MSG, PASSWORD_BAD_MIN_LENGTH_MSG, PASSWORD_BAD_MAX_LENGTH_MSG,
+    REQUIRED_FIELD_COUNTRY_MSG, REQUIRED_FIELD_CONFIRM_EMAIL_MSG, REQUIRED_FIELD_NAME_MSG
```
I can do that, yeah, but I'm just worried that the resulting column would be so huge that it'd become the biggest existing in the codebase! I think it would be a more reasonable move to do `accounts.X` instead, though, because then we wouldn't have to deal with this mess of a list for adding/deleting things (and I've had to do that several times, and it wasn't fun).

UmanShahzad commented on openedx/core/djangoapps/user_api/accounts/api.py:
```diff
@@ -20,16 +21,24 @@
+    AccountUpdateError, AccountValidationError,
+    AccountDataBadLength, AccountDataBadType,
+    AccountUsernameInvalid, AccountPasswordInvalid, AccountEmailInvalid, AccountCountryInvalid,
+    AccountUserAlreadyExists, AccountUsernameAlreadyExists, AccountEmailAlreadyExists,
```
Will try to just refer to them as `errors.X` or something.

UmanShahzad commented on openedx/core/djangoapps/user_api/views.py:
```diff
@@ -29,13 +29,12 @@
+    PASSWORD_MAX_LENGTH, PASSWORD_MIN_LENGTH,
+    USERNAME_MAX_LENGTH, USERNAME_MIN_LENGTH,
+    EMAIL_CONFLICT_MSG, USERNAME_CONFLICT_MSG,
+    REQUIRED_FIELD_CONFIRM_EMAIL_MSG, REQUIRED_FIELD_COUNTRY_MSG
```
Will try to refer to them with `accounts.X`.

UmanShahzad commented:
@tikr This was previously rebased on `master` and the hash of the tip of this branch before that was [`1fbd144`](https://github.com/open-craft/edx-platform/commit/1fbd144ef68b4a760f31bfbbe291df20fa846885).

@haikuginger Thanks for the review and good comments! I've replied to all comments and am going to be making the (small) changes, and will ping you again for your approval.

@brittneyexline / @douglashall If any of you would like to also review that'd be great -- and if an accessibility / UI review is possible before the end of this week so we can get this over the line, this would be ready for merge.

UmanShahzad commented:
@haikuginger I forgot to address your concern about certain fields not being validated -- so yes, I haven't bothered to deal with those yet. What I'm thinking is to implement all handlers for all existing registration fields, and just go from there. An alternative is to just hack up a solution where if the field is required but not in our list of available live validation fields, then do as you said: if the field is empty/unchecked/unselected, give it a generic error response (BTW, these error responses can be found using `$el.data('errormsg-required')`). What route should we go? I'm currently thinking the latter, because implementing a backend for all fields can take a while.

haikuginger commented:
@UmanShahzad, I agree - let's go with the latter method.

UmanShahzad commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -32,10 +32,11 @@
+        >
+            <% _.each(options, function(el) { %>
+                <option value="<%- el.value%>"<% if ( el.default ) { %> data-isdefault="true" selected<% } %>><%- el.name %></option>
+            <% }); %>
```
There's no actual changes here: formatting nits because this was getting incredibly hard to read when I had to read it.

UmanShahzad commented on lms/templates/student_account/form_field.underscore:
```diff
@@ -59,14 +60,15 @@
+            <div class="supplemental-link">
+                <a href="<%- supplementalLink %>" target="_blank"><%- supplementalText %></a>
+            </div>
+        <% } %>
```
Same here.

UmanShahzad commented:
@haikuginger Ready for your review :) This should now have generic validation for all fields for which there isn't a backend handler. 

NOTE: I'm currently not letting out error messages for `gender` and `year_of_birth` fields, because they cause some nasty UI issues for which I'm not sure which direction to take. The issue is, when one of those fields displays an error message and the other doesn't, the one that has no error message gets misaligned with the one that has the error message, because one is taking up more height than the other.

There are many solutions to this but each would require extensive UI review and I'd like to get @chris-mike 's thoughts on whether we should even bother or what solution to try. A hacky solution would be to put those 2 fields specifically together into a div, for example, and another would be to make them take the full width as the other fields do, and so on.

UmanShahzad commented:
jenkins run all

UmanShahzad commented:
I seem to have lost access to jenkins?

haikuginger commented:
jenkins run all

UmanShahzad commented:
@haikuginger Thanks :) Just waiting on an accessibility review from @cptvitamin and hopefully 1 more UI review from @chris-mike if possible? Then I'll squash.

UmanShahzad commented:
I've squashed this on [`b118ebe`](https://github.com/open-craft/edx-platform/commit/b118ebe0d5be3441dc6160cee1fede294a588d82) to make it easier to cherry-pick onto the business sandbox, and because Jesse already reviewed the code. Only a UI/Accessibility review is in order.

UmanShahzad commented:
The quality failure is a super small nit.. I'll push the 1-line change later.

UmanShahzad commented:
Addressed the accessibility review (@cptvitamin -- double check if needed) & rebased. Old tip hash was [6ba4122](https://github.com/open-craft/edx-platform/commit/6ba4122aba61b7c40b677c6fbc9c1db1b25681b4). I'll update the sandbox next.

@cptvitamin @douglashall @brittneyexline @itsjeyd @haikuginger Good to look at merging this by Monday or so?

cptvitamin commented:
👍

UmanShahzad commented:
Rebased on [`ccaaa43`](https://github.com/open-craft/edx-platform/commit/ccaaa43574741204ea7590f24a40ce3d38ab03e1).

@brittneyexline Thanks :) I'm available for answering any questions / issues that pop up during the week that this is merged. I've addressed the one nit Christine mentioned as well (just had to remove a newline).

I've fully squashed things now to 2 commits (backend, frontend).

UmanShahzad commented:
I've updated the main PR post with the most recent screenshots from the business sandbox.

UmanShahzad commented:
The remaining broken tests are project-wide afaik.

UmanShahzad commented:
jenkins run bokchoy

UmanShahzad commented:
@haikuginger @itsjeyd @douglashall @brittneyexline @chris-mike @cptvitamin 
Thanks a lot for working on getting this in!

itsjeyd commented:
🎉
